<a href="https://colab.research.google.com/github/shanshanli806/Artificial-Intelligence/blob/main/Assignment04/CSc448_A04.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd

## Part 1: Load the dataset

In [2]:
from sklearn.datasets import load_iris

# Load the dataset (load remotely, not locally)
iris = load_iris()
iris_df = pd.DataFrame(iris.data, columns=iris.feature_names)
iris_df['species'] = iris.target

In [3]:
# Output the first 15 rows of the data
iris_df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [4]:
# Display a summary of the table information (number of datapoints, etc.)
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


In [5]:
iris_df.shape # (rows, colums)

(150, 5)

In [6]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [7]:
# printing the name of the different species
print("Target: ", iris.target_names)
iris_df['species'].value_counts()

Target:  ['setosa' 'versicolor' 'virginica']


0    50
1    50
2    50
Name: species, dtype: int64

### About the dataset

<sub>Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?</sub>

The Iris Flower dataset is a multivariable dataset that includes 150 records of the four features measured from the 50 samples of each of the three species of Iris.

The **attributes** of each record:
   1. **sepal length (cm)**
   2. **sepal width (cm)**
   3. **petal length (cm)**
   4. **petal width (cm)**
   5. **species** with the values 0,1,2 corresponding to the three different species.
       * 0 = setosa, 
       * 1 = versicolor, 
       * 3 = virginica

## Part 2: Spli the dataset into train and test

In [8]:
# Printing the columns in the dataset
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'species'],
      dtype='object')

In [9]:
# Take the pandas dataset and split it into our features (X) and label (y)
# features (X)
features = iris_df[['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']]
# label (y)
label = iris_df['species']

In [10]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features, label, train_size=0.9, test_size=0.1)

In [11]:
X_train.shape, y_train.shape

((135, 4), (135,))

In [12]:
X_test.shape, y_test.shape

((15, 4), (15,))

## Part 3: Logistic Regression

In [13]:
# i. Use sklearn to train a LogisticRegression model on the training set
from sklearn.linear_model import LogisticRegression

LR = LogisticRegression(max_iter = 500)
LR.fit(X_train, y_train) # training the model

LogisticRegression(max_iter=500)

In [14]:
# ii. For a sample datapoint, predict the probabilities for each possible class
LR_pred = LR.predict_proba(X_test[:1])
pd.DataFrame(LR_pred, columns=['setosa', 'versicolor', 'virginica'])

setosa  versicolor     virginica
0  0.988292    0.011708  1.094564e-08

In [15]:
# iii. Report on the score for Logistic regression model, what does the score measure?
score =LR.score(X_test, y_test)
print("The score for the model is", score)

The score for the model is 0.8666666666666667


`LR.score(X_test, y_test)` is the prediction for `X_test` and compares it with `y_test`, and returns the accuracy score of the model.

In [16]:
print("Here, the accuracy for this model is {0:.2%}".format(score))

Here, the accuracy for this model is 86.67%


In [17]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
# coefficents
coef = LR.coef_
# intercepts
intercept = LR.intercept_

LR_df = pd.DataFrame(coef, iris.target_names, columns=iris.feature_names)
LR_df['intercept'] = intercept
LR_df

sepal length (cm)  sepal width (cm)  petal length (cm)  \
setosa              -0.435366          0.901588          -2.440397   
versicolor           0.395122         -0.475102          -0.196183   
virginica            0.040244         -0.426485           2.636581   

            petal width (cm)  intercept  
setosa             -1.033326   9.815816  
versicolor         -0.913030   3.428572  
virginica           1.946355 -13.244388

## Part 4: Support Vector Machine

In [18]:
# i. Use sklearn to train a Support Vector Classifier on the training set
from sklearn import svm

SVM_model = svm.SVC(probability = True)
SVM_model.fit(X_train, y_train) # training the model

SVC(probability=True)

In [19]:
# ii. For a sample datapoint, predict the probabilities for each possible class
SVM_pred = SVM_model.predict_proba(X_test[:1])
pd.DataFrame(SVM_pred, columns=['setosa', 'versicolor', 'virginica'])

setosa  versicolor  virginica
0  0.977412    0.011616   0.010972

In [20]:
# iii. Report on the score for the SVM, what does the score measure?
SVM_score = SVM_model.score(X_test, y_test)
print("The score for the model is", SVM_score)

The score for the model is 0.9333333333333333


`svm_model.score(X_test, y_test)` is the prediction for `X_test` and compares it with `y_test`, and returns the mean accuracy score of the model.

* `X_test` is the test data/samples.
* `y_test` is the labels (setosa, versicolor, virginica)

In [21]:
print("Here, the accuracy for this model is {0:.2%}".format(SVM_score))

Here, the accuracy for this model is 93.33%


## Part 5: Neural Network

In [22]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
from sklearn.neural_network import MLPClassifier

NN = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1)
NN.fit(features, label)

MLPClassifier(alpha=1e-05, hidden_layer_sizes=(15,), random_state=1,
              solver='lbfgs')

In [23]:
# ii. For a sample datapoint, predict the probabilities for each possible class
NN_pred = NN.predict_proba(X_test[:1])
pd.DataFrame(NN_pred, columns=['setosa', 'versicolor', 'virginica'])

setosa     versicolor      virginica
0     1.0  5.592259e-169  1.812251e-198

In [24]:
# iii. Report on the score for the Neural Network, what does the score measure?
NN_score = NN.score(X_test, y_test)
print("The score for the model is", NN_score)

The score for the model is 1.0


In [25]:
print("Here, the accuracy for this model is {0:.2%}".format(NN_score))

Here, the accuracy for this model is 100.00%


`NN.score(X_test, y_test)` is the prediction for `X_test` and compares it with `y_test`, and returns the mean accuracy score of the model.

* `X_test` is the test data/samples.
* `y_test` is the labels (setosa, versicolor, virginica)

In [26]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
NN_exp = MLPClassifier(max_iter=500, learning_rate='constant', solver='adam', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1)
NN_exp.fit(X_train, y_train)

C:\Users\shans\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(alpha=1e-05, hidden_layer_sizes=(15,), max_iter=500,
              random_state=1)

In [27]:
# ii. For a sample datapoint, predict the probabilities for each possible class
NN_pred_new = NN_exp.predict_proba(X_test[:1])
pd.DataFrame(NN_pred_new, columns=['setosa', 'versicolor', 'virginica'])

setosa  versicolor  virginica
0  0.987101    0.005505   0.007393

In [28]:
# iii. Report on the score for the Neural Network, what does the score measure?
NN_score_new = NN_exp.score(X_test, y_test)
print("The score for the model is", NN_score_new)

The score for the model is 0.8666666666666667


In [29]:
print("Here, the accuracy for this model is {0:.2%}".format(NN_score_new))

Here, the accuracy for this model is 86.67%


## Part 6: K-Nearest Neighbors

In [30]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

from sklearn.neighbors import KNeighborsClassifier

KNN = KNeighborsClassifier(n_neighbors=4)
KNN.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=4)

In [31]:
# ii. For a sample datapoint, predict the probabilities for each possible class
KNN_pred = KNN.predict_proba(X_test[:1])
pd.DataFrame(KNN_pred, columns=['setosa', 'versicolor', 'virginica'])

setosa  versicolor  virginica
0     1.0         0.0        0.0

In [32]:
# iii. Report on the score for kNN, what does the score measure?
KNN_score = KNN.score(X_test, y_test)
print("The score for the model is", KNN_score)
print("Here, the accuracy for this model is {0:.2%}".format(KNN_score))

The score for the model is 0.8666666666666667
Here, the accuracy for this model is 86.67%


## Part 7: Conclusions and takeaways

<sub>In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?</sub>

* **Logistic Regression**
    * score: 86.67%
* **Support Vector Machine**
    * score: 93.33%
* **Neural Network**
    * MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1)
        * score: 100.00%
    * MLPClassifier(max_iter=500, learning_rate='constant', solver='adam', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)
        * score: 86.67%
* **K-Nearest Neighbors**
    * score: 86.67%

The Neural Network model score the highest. However, the Neural Network model scored differently with different options. Also, it is to say that each time I restart and run all the models again, all the models scores different in an accuracy between 80% and 100%. This might be due to the different split we get each time we rerun `X_train, X_test, y_train, y_test = train_test_split(features, label, train_size=0.9, test_size=0.1)`.